application/client.core/src/app/environment/components/sidebar/shell/environment/environment_variables/component.ts

```diff
@@ -0,0 +1,198 @@
+
+@Component({
+    selector: 'app-sidebar-app-shell-environment-variables',
+    templateUrl: './template.html',
+    styleUrls: ['./styles.less']
+})
+
+export class SidebarAppShellEnvironmentVariablesComponent implements OnInit, OnDestroy {
+
+    @Input() public information: INewInformation;
+    @Input() public setEnvironment: (information: INewInformation) => void;
+    @Input() public close: () => {};
+
+    @ViewChild('variableAdd') _ng_variableAdd: ElementRef<HTMLInputElement>;
+
+    public readonly _ng_variable: string = 'variable';
+    public readonly _ng_value: string = 'value';
+    public _ng_selected: IEnvironment;
+    public _ng_adding: boolean = false;
+    public _ng_height: number = 200;
+    public _ng_newVariable: string = '';
+    public _ng_newValue: string = '';
+
+    private _prevValue: string = '';
+    private _prevVariable: string = '';
+    private _editing: IEnvironment | undefined;
+
+    constructor() { }
+
+    public ngOnInit() {
+        this._sortInformation();
+        this._ng_height = Math.round(window.innerHeight * .7);
+    }
+
+    public ngOnDestroy() {
+        if (this._ng_selected !== undefined) {
+            this._ng_selected.editing = {
+                value: false,
+                variable: false
+            };
+            this._ng_selected = undefined;
+        }
+        this._ng_adding = false;
+        this.setEnvironment(this.information);
+    }
+
+    public _ng_onContexMenu(event: MouseEvent, env: IEnvironment, type: string) {
+        const items: IMenuItem[] = [
+            {
+                caption: 'Edit',
+                handler: () => {
+                    this._changeFocus(env);
+                    if (type === this._ng_value) {
+                        env.editing.value = true;
+                        this._prevValue = env.value;
+                    } else {
+                        env.editing.variable = true;
+                        this._prevVariable = env.variable;
+                    }
+                }
+            },
+            {
+                caption: 'Remove',
+                handler: () => {
+                    this._changeFocus();
+                    this.information.env = this.information.env.filter((ngEnv: IEnvironment) => {
+                        return ngEnv.variable !== env.variable && ngEnv.value !== env.value;
+                    });
+                },
+            },
+        ];
+        ContextMenuService.show({
+            items: items,
+            x: event.pageX,
+            y: event.pageY,
+        });
+    }
+
+    public _ng_onKeyUp(event: KeyboardEvent) {
+        if (event.key !== 'Enter') {
+            return;
+        }
+        this._finishCreating(true);
+        this._setCustomStatus();
+        this._sortInformation();
+    }
+
+    public _ng_onClick(env: IEnvironment) {
+        if (this._ng_selected !== env) {
+            this._changeFocus();
+        }
```

Comment:
Compering of objects isn't a trustable thing in JS. Maybe you want to compare by some `id/uuid`? Like
```
if (this._ng_selected.uuid !== env.uuid) { ... }
```
As soon as name of variable is unique in scope of envvars, property `variable` can be used as unique id:
```
if (this._ng_selected.variable !== env.variable) { ... }
```